lapack/testlapack/dgeqp3.go

```diff
@@ -20,120 +22,95 @@ type Dgeqp3er interface {
-				}
-			}
-			// Allocate a slice for scalar factors of elementary
-			// reflectors and fill it with random numbers. Dgeqp3
-			// will overwrite them with valid data.
-			k := min(m, n)
-			tau := make([]float64, k)
-			for i := range tau {
-				tau[i] = rnd.Float64()
-			}
-			// Get optimal workspace size for Dgeqp3.
-			work := make([]float64, 1)
-			impl.Dgeqp3(m, n, a, lda, jpvt, tau, work, -1)
-			lwork := int(work[0])
-			work = make([]float64, lwork)
-			for i := range work {
-				work[i] = rnd.Float64()
+	for _, m := range []int{0, 1, 2, 3, 4, 5, 12, 23, 129} {
+		for _, n := range []int{0, 1, 2, 3, 4, 5, 12, 23, 129} {
+			for _, lda := range []int{max(1, n), n + 3} {
+				dgeqp3Test(t, impl, rnd, m, n, lda)
 			}
+		}
+	}
+}
 
-			// Compute a QR factorization of A with column pivoting.
-			impl.Dgeqp3(m, n, a, lda, jpvt, tau, work, lwork)
+func dgeqp3Test(t *testing.T, impl Dgeqp3er, rnd *rand.Rand, m, n, lda int) {
+	const (
+		tol = 1e-14
 
-			// Compute Q based on the elementary reflectors stored in A.
-			q := constructQ("QR", m, n, a, lda, tau)
-			// Check that Q is orthogonal.
-			if !isOrthogonal(q) {
-				t.Errorf("Case %v, Q not orthogonal", c)
-			}
+		all = iota
+		some
+		none
+	)
+	for _, free := range []int{all, some, none} {
+		name := fmt.Sprintf("m=%d,n=%d,lda=%d,", m, n, lda)
 
-			// Copy the upper triangle of A into R.
-			r := blas64.General{
-				Rows:   m,
-				Cols:   n,
-				Stride: n,
-				Data:   make([]float64, m*n),
-			}
-			for i := 0; i < m; i++ {
-				for j := i; j < n; j++ {
-					r.Data[i*n+j] = a[i*lda+j]
-				}
+		// Allocate m×n matrix A and fill it with random numbers.
+		a := randomGeneral(m, n, lda, rnd)
+		// Store a copy of A for later comparison.
+		aCopy := cloneGeneral(a)
+		// Allocate a slice of column pivots.
+		jpvt := make([]int, n)
+		for j := range jpvt {
+			switch free {
+			case all:
+				// All columns are free.
+				jpvt[j] = -1
+				name += "free=all"
+			case some:
+				// Some columns are free, some are leading columns.
+				jpvt[j] = rnd.Intn(2) - 1 // -1 or 0
+				name += "free=some"
+			case none:
+				// All columns are leading.
+				jpvt[j] = 0
+				name += "free=none"
+			default:
+				panic("bad freedom")
 			}
-			// Compute Q * R.
-			got := nanGeneral(m, n, lda)
-			blas64.Gemm(blas.NoTrans, blas.NoTrans, 1, q, r, 0, got)
-			// Compute A * P: rearrange the columns of A based on the permutation in jpvt.
-			want := blas64.General{Rows: m, Cols: n, Stride: lda, Data: aCopy}
-			impl.Dlapmt(true, want.Rows, want.Cols, want.Data, want.Stride, jpvt)
-			// Check that A * P = Q * R.
-			if !equalApproxGeneral(got, want, 1e-13) {
-				t.Errorf("Case %v,  Q*R != A*P\nQ*R=%v\nA*P=%v", c, got, want)
+		}
+		// Allocate a slice for scalar factors of elementary
+		// reflectors and fill it with random numbers. Dgeqp3
+		// will overwrite them with valid data.
+		k := min(m, n)
+		tau := make([]float64, k)
+		for i := range tau {
+			tau[i] = rnd.Float64()
```

Comment:
Just a thought, I think pretty much all of our random tests use `rnd.Float64()` values, which means they are in [0,1). Is that a problem?

Reply:
In tests that I write I use NormFloat64 but in general I don't think it's a problem, it fulfils the purpose of testing.